File: src/taeeun/week5/자물쇠와 열쇠.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;
typedef vector<vector<int>> vi;
int n, m;

//key를 회전시키는 함수
vi rotate(vi &key) {
    vi copy(m, vector<int>(m));
    for (int i=0; i<m; i++) {
        for (int j=0; j<m; j++) {
            copy[i][j]=key[m-j-1][i];
        }
    }
    return copy;
}

//열쇠와 자물쇠를 합치는 함수
vi combine(vi &key, vi &lock, int x, int y) {
    //너비가 m+n+m인 사각형이 나온다
    vi board(n+2*m, vector<int>(n+2*m));
    //lock의 위치는 고정
    for (int i=m; i<n+m; i++) {
        for (int j=m; j<n+m; j++) {
            board[i][j]=lock[i-m][j-m];
        }
    }
    //key는 (x,y)만큼 이동
    for (int i=x; i<x+m; i++) {
        for (int j=y; j<y+m; j++) {
            board[i][j]+=key[i-x][j-y];
        }
    }
    
    return board;
}

bool check(vi &key, vi&lock, int x, int y) {
    vi board=combine(key, lock, x, y);
    for (int i=m; i<n+m; i++) {
        for (int j=m; j<n+m; j++) {
            if (board[i][j]!=1) return false;
        }
    }
    return true;
}

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    bool answer = true;
    m=key.size();
    n=lock.size();
    
    //key가 lock 상하좌우로 붙을 수 있음
    for (int i=0; i<n+m; i++) {
        for (int j=0; j<n+m; j++) {
            //4번 돌려볼 수 있음
            for (int k=0; k<4; k++) {
                key=rotate(key);
                if (check(key, lock,i, j)) return true;
            }
        }
    }
    return false;
}

File: src/chanhyun/week5/자물쇠와 열쇠.py
def rotate(key):
    M = len(key)
    newKey = [[0 for _ in range(M)] for _ in range(M)]
    for i in range(M):
        for j in range(M):
            newKey[M-1-j][i] = key[i][j]
    return newKey

def check(board):
    N = len(board)//3
    for i in range(N, N*2):
        for j in range(N, N*2):
            if board[i][j] != 1:
                return False
    return True

def solution(key, lock):
    N = len(lock)
    M = len(key)
    board = [[0 for _ in range(N*3)] for _ in range(N*3)]
    for i in range(N):
        for j in range(N):
            board[N+i][N+j] = lock[i][j]
    for _ in range(4):
        key = rotate(key)
        for i in range(N*2):
            for j in range(N*2):
                for x in range(M):
                    for y in range(M):
                        board[i+x][j+y] += key[x][y]
                if check(board):
                    return True
                for x in range(M):
                    for y in range(M):
                        board[i+x][j+y] -= key[x][y]
    return False